graphql_compiler/compiler/emit_sql.py

```diff
@@ -1014,10 +1016,23 @@ def recurse(self, vertex_field: str, depth: int) -> None:
         used_columns = sorted(self._used_columns[self._current_location.query_path])
 
         # The base of the recursive CTE selects all needed columns and sets the depth to 0
+        base_alias = self._current_alias.alias()
         base = sqlalchemy.select(
-            [previous_alias.c[col].label(col) for col in used_columns]
-            + [previous_alias.c[primary_key].label(CTE_KEY_NAME), literal_0.label(CTE_DEPTH_NAME),]
-        ).cte(recursive=True)
+            [base_alias.c[col].label(col) for col in used_columns]
+            + [base_alias.c[primary_key].label(CTE_KEY_NAME), literal_0.label(CTE_DEPTH_NAME),]
+        )
+        if self._recurse_needs_cte:
+            # Optimization: Only compute the recursion for the valid starting points -- ones that
+            # will not be discarded when the recursive cte is joined to the rest of the query.
+            # The SQL database could do this predicate pushdown itself, but most implementations
+            # don't.
+            # This optimization is absolutely necessary for queries with strong filters before
```

Comment:
```suggestion
            # This optimization is absolutely necessary for queries with highly selective filters before
```
and this might force re-wrapping of lines to fit within 100 chars, sorry 😬